cf-custom-resources/lib/dns-cert-validator.js

```diff
@@ -238,104 +238,74 @@ const updateHostedZoneRecords = async function (
-// Therefore, we check if there is more than one certificate using the record and only delete
-// if there is no other certificate using the record.
 const deleteHostedZoneRecords = async function (
   options,
-  certDomain,
+  certArn,
+  defaultDomain,
   envRoute53,
   appRoute53,
   acm,
   envHostedZoneId
 ) {
-  let isLegacyCert = false;
-  // Legacy cert only has two DomainValidationOptions:
-  // `${envName}.${appName}.${domainName}` and `*.${envName}.${appName}.${domainName}`
-  if (options.length <= 2) {
-    isLegacyCert = true;
-  }
-
-  const certsWithEnvDomain = await numOfGeneratedCertificates(acm, certDomain);
-  const isLastOne = certsWithEnvDomain === 1;
-
-  const newOptions = [];
-  switch (`${isLegacyCert}|${isLastOne}`) {
-    case `true|true`:
-      // If it is a legacy cert and it is the last Copilot cert,
-      // we'll go ahead to remove the validation CNAME record in env hosted zone.
-      newOptions.push(...options);
-      break;
-    case `true|false`:
-      // If it is a legacy cert but it is not the last Copilot cert,
-      // we'll do nothing since the new Copilot cert needs validation CNAME record
-      // in the env hosted zone.
-      break;
-    case `false|true`:
-      // If it is not a legacy cert and it is the last Copilot cert,
-      // we'll remove all validation CNAME records in env/app/root hosted zone.
-      newOptions.push(...options);
-      break;
-    case `false|false`:
-      // If it is not a legacy cert and it is not the last Copilot cert,
-      // we'll remove validation CNAME records only for app and root hosted zone,
-      // since the legacy cert still needs the validation record in the env hosted zone.
-      for (const option of options) {
-        if (option.DomainName === certDomain || option.DomainName === `*.${certDomain}`) {
-          continue;
-        }
-        newOptions.push(option);
+  let listCertificatesInput = {};
+  let newCertOptions = [];
+  // Make sure if there is a new certificate, we don't delete any DNS validation records
+  // that are used by the new certificate.
+  while (true) {
+    const listCertResp = await acm
+      .listCertificates(listCertificatesInput)
+      .promise();
+    for (const certSummary of listCertResp.CertificateSummaryList || []) {
+      if (
+        certSummary.DomainName !== defaultDomain ||
+        certSummary.CertificateArn === certArn
+      ) {
+        // Skip if it is not the new certificate.
+        continue;
       }
+      const { Certificate } = await acm
+        .describeCertificate({
+          CertificateArn: certSummary.CertificateArn,
+        })
+        .promise();
+      newCertOptions = Certificate.DomainValidationOptions || [];
       break;
+    }
+    if (!listCertResp.NextToken) {
+      break;
+    }
+    listCertificatesInput.NextToken = listCertResp.NextToken;
+  }
+  const newCertOptionSet = new Set(
+    newCertOptions.map((item) => item.DomainName)
+  );
```

Comment:
Can we rename these variables like this? 
nit: personally I find it harder to read once we break it up into these new lines
```suggestion
  const newCertSANS = new Set(newCertOptions.map((item) => item.DomainName));
```

Reply:
It's formatted by vscode 🤔 I'll rename the variables.

Reply:
💩

Reply:
ok after renaming it collapses to one line